### PR TITLE
String::remove_matches O(n^2) -> O(n)

### PR DESCRIPTION
**Pull request by tamird**
Copy only non-matching bytes. Replace collection of matches into a
vector with iteration over rejections, exploiting the guarantee that we
mutate parts of the haystack that have already been searched over.

r? @joshtriplett

### CODE REVIEW

**Comment by tamird** [comment]
cc @jcotton42 @pickfire

**Comment by jcotton42** [comment]
Clever. Not sure why I didn't think of that.

File: library/alloc/src/string.rs [hunk 1]
```diff
@@ -1246,11 +1240,11 @@ impl String {
         // SAFETY: start and end will be on utf8 byte boundaries per
         // the Searcher docs
         unsafe {
-            for (start, end) in matches {
-                ptr::copy(
-                    self.vec.as_mut_ptr().add(end - shrunk_by),
-                    self.vec.as_mut_ptr().add(start - shrunk_by),
-                    len - end,
+            for (start, end) in matches.into_iter().rev() {
+                let () = ptr::copy(
```
Comment (jcotton42) [inline library/alloc/src/string.rs:1244]: Why is this `let ()` here?

Comment (tamird) [reply library/alloc/src/string.rs:1244]: Style thing. This lets the reader know that there's no return value being discarded.

We use Rust pretty heavily in Fuchsia, and learned that all too often implicit drop can lead to `Result`s being dropped on the floor - so we always destructure on the LHS.

File: library/alloc/src/string.rs [hunk 2]
```diff
@@ -1227,17 +1227,11 @@ impl String {
     where
         P: for<'x> Pattern<'x>,
     {
-        use core::str::pattern::Searcher;
+        use core::str::pattern::Searcher as _;
```
Comment (pickfire) [inline library/alloc/src/string.rs:1230]: Is this necessary?

Comment (tamird) [reply library/alloc/src/string.rs:1230]: No. Style choice - brings in the trait methods without bringing in the `Searcher` item.

File: library/alloc/src/string.rs (hunk 1 above)
Comment (pickfire) [inline library/alloc/src/string.rs:1244]: Why do we need `let () = `?

**Review by pickfire — APPROVED** [review]
Other than the mentioned comments it looks good to me but not sure if it improves the performance. Maybe we should do a perf run?

File: library/alloc/src/string.rs (hunk 1 above)
Comment (tamird) [reply library/alloc/src/string.rs:1244]: There's already a thread about this.

Comment (pickfire) [reply library/alloc/src/string.rs:1244]: Oh, I left the same comment. I think we should keep it similar to how rust handles it. Maybe fuchsia does but rust library `ptr::copy` does not do implicit drop. Also, `Result` have `must_use` and cannot be dropped like that.

File: library/alloc/src/string.rs (hunk 2 above)
Comment (tamird) [reply library/alloc/src/string.rs]: Reverted.

File: library/alloc/src/string.rs (hunk 1 above)
Comment (tamird) [reply library/alloc/src/string.rs]: Reverted.

**Comment by tamird** [comment]
Not sure what a perf run would do - does this function have good benchmark coverage? Seems like no and since it's new it's unlikely to show up in any other perf paths.

**Comment by slightlyoutofphase** [comment]
> Other than the mentioned comments it looks good to me but not sure if it improves the performance. Maybe we should do a perf run?

To me it seems like the general idea of reversing the process would be likely to improve performance, but the overall indirection of going through `FromFn` and then the `Vec::from_iter` implementation instead of just pushing, as well as the indirection of needing to turn the Vec into an `IntoIter` later on might possibly in turn offset the performance gains negatively.

**Comment by pickfire** [comment]
> Not sure what a perf run would do - does this function have good benchmark coverage? Seems like no and since it's new it's unlikely to show up in any other perf paths.

No, it's not about benchmark. Although it does have benchmarks but perf run is to check if this affects compile time and runtime of generated code.

**Comment by m-ou-se** [comment]
After this change, all bytes are still copied/moved several times (exactly the number of matches before them). With 100 matches, the last segment gets moved 100 times.

You cannot efficiently do this in reverse: The last segment cannot be put into the right place right away, because its new place still holds data that needs to be moved. Doing this in order starting at the first match means that every segment can directly be moved into the right position. To do so, you'd have to look at the start of the next match to find the end of the segment to move.

**Comment by tamird** [comment]
@m-ou-se there's a kernel of truth to what you say, but the optimization you describe requires bookkeeping which does not exist before this change.

Consider this string:

```
[-----|match1|-----|match2|-----|match3|----|matchN|---]
```

When iterating in forward-order, all bytes after `match1` will be copied, then all bytes after `match2` will be copied, and so on. In order to do what you say, the implementation would have to defer copying the non-matching interstitial (e.g. between match1 and match2) until after the *next* match is found. I think it would be possible to implement that, but it's not in place today. Thoughts?

**Comment by m-ou-se** [comment]
@tamird Yes, exactly. The current implementation and your implementation are both O(n²). Iterating forwards allows for an O(n) implementation, by only copying the part until the next match.

> would have to defer copying the non-matching interstitial (e.g. between match1 and match2) until after the next match is found. I think it would be possible to implement that, but it's not in place today.

The current implementation already finds all the matches and puts them all in a Vec before it moves anything, so all the information is already there.

**Comment by tamird** [comment]
@m-ou-se how does this look?

**Comment by m-ou-se** [comment]
Very nice!

It'd be good to narrow the scope of the `unsafe` block by putting `unsafe { .. }` only around the copy_nonoverlapping and set_len lines, each with their own `// SAFETY:` comment.

**Comment by tamird** [comment]
Done. Replaced `next_match` with `next_reject` to simplify things and also removed batching up matches into a vector made possible by that change.

**Comment by tamird** [comment]
@m-ou-se ping.

**Comment by m-ou-se** [comment]
> also removed batching up matches into a vector made possible by that change

Unfortunately, that's technically not allowed because the Searcher owns the entire &str, even the parts that it already processed. (That might be used by e.g. a regular expression with backreferences.) So, that `copy_nonoverlapping` is modifying data that is immutably borrowed. :(

It's a bit of a shame, because it actually works 'fine' for all of the implementations of Searcher we have right now in the standard library.

An alternative to the Vec might be to not keep a single Searcher around, but instead create a new Searcher every iteration for finding the next reject/match. (Or maybe we need a better Pattern/Searcher API, but that's a much bigger challenge.)

**Comment by tamird** [comment]
Ack, resurrected the vector.

**Comment by m-ou-se** [comment]
Looks like using `next_reject` instead of `next_match` can be very inefficient, unfortunately. The `Searcher` documentation says:

> A Match result needs to contain the whole matched pattern, however Reject results may be split up into arbitrary many adjacent fragments. Both ranges may have zero length.

Looking at the implementation, the searchers have an efficient `next_match` implementation, while `next_reject` returns as quickly as possible:

```rust
#![feature(pattern)]

use std::str::pattern::{Searcher, Pattern};

fn main() {
    let s = "aaaaa_aaaaa_aaaa";
    let mut searcher = '_'.into_searcher(s);
    while let Some((a, b)) = searcher.next_reject() {
        println!("reject: {:?}: {:?}", a..b, &s[a..b]);
    }
}
```

```
reject: 0..1: "a"
reject: 1..2: "a"
reject: 2..3: "a"
reject: 3..4: "a"
reject: 4..5: "a"
reject: 6..7: "a"
reject: 7..8: "a"
reject: 8..9: "a"
reject: 9..10: "a"
reject: 10..11: "a"
reject: 12..13: "a"
reject: 13..14: "a"
reject: 14..15: "a"
reject: 15..16: "a"
```

**Comment by tamird** [comment]
Life is so hard. Added some code to coalesce rejection regions.

**Comment by m-ou-se** [comment]
Why not just go back to the `next_match` version? That's probably both simpler and more efficient. This version still ends up calling `next_reject` for every single character (for most Searchers), while the searchers have an efficient `next_match` implementation.

**Comment by tamird** [comment]
Can you help me understand why you think it's more efficient? I don't think it's simpler - it requires synthetic start and end "matches"; the real thing we want here is to preserve all rejections, which feels more obvious to me when using `next_reject`.

**Comment by m-ou-se** [comment]
> Can you help me understand why you think it's more efficient?

For the searchers we have (e.g. `CharSearcher`), `next` will return every time it decided the next character is a reject or a match. So for a string of a 1000 characters, that might take up to 1000 iterations. These searchers have a more efficient `next_match` implementation, which will use e.g. `memchr` to directly skip over to the next match. `memchr` is a lot faster than iterating over the chars individually and checking each of them (combined with code that then coalesces the results).

Example of one of those implementations: https://github.com/rust-lang/rust/blob/2f601ef527a37de9fb181c5fb1a6c06d06218ed3/library/core/src/str/pattern.rs#L370#

> the real thing we want here is to preserve all rejections, which feels more obvious to me when using next_reject.

You could also say we want to remove all the matches, making `next_match` the 'obvious' choice. ^^

Regardless, I think this all shows that the unstable Pattern could use some redesigning. Luckily it's still unstable. It's far from obvious what a better design would be though.

**Comment by tamird** [comment]
OK, back to using `next_match`! PTAL.

**Comment by m-ou-se** [comment]
Thanks a lot!

@bors r+

**Comment by klensy** [comment]
Any bench results to see the difference?

**Comment by m-ou-se** [comment]
It changes quadratic time to linear time, so the differences can get huge:

```rust
#![feature(string_remove_matches)]

use std::time::Instant;

fn main() {
    let start = Instant::now();
    let mut s = ("a".repeat(5000) + "b").repeat(5000);
    s.remove_matches('b');
    println!("{:?}", start.elapsed());
}
```

Before: 7.123511259s
After: 15.944429ms

File: library/alloc/src/string.rs [hunk 3]
```diff
@@ -1290,32 +1290,49 @@ impl String {
...
+            let count = end - start;
+            if start != len {
+                // SAFETY: per Searcher::next:
+                //
+                // The stream of Match and Reject values up to a Done will
+                // contain index ranges that are adjacent, non-overlapping,
+                // covering the whole haystack, and laying on utf8
+                // boundaries.
+                unsafe {
+                    ptr::copy_nonoverlapping(ptr.add(start), ptr.add(len), count);
```
Comment (tmiasko) [inline library/alloc/src/string.rs:1328]: The copy source and destination are overlapping.

Comment (tamird) [reply library/alloc/src/string.rs:1328]: Oops, you're right. Done.

Comment (m-ou-se) [reply library/alloc/src/string.rs]: Oh oops, good catch :)

**Comment by JohnTitor** [comment]
Should be perf-sensitive, @bors rollup=never

**Comment by m-ou-se** [comment]
> Should be perf-sensitive, @bors rollup=never

This function isn't used anywhere in the standard library or compiler.

**Comment by JohnTitor** [comment]
> This function isn't used anywhere in the standard library or compiler.

Oh sorry, my bad!

**Pull request merged**